lorawan-macros/src/lib.rs

```diff
@@ -0,0 +1,449 @@
+
+    let mut payload_struct_impls = Vec::new();
+
+    let mut payload_struct_creator_impls = Vec::new();
+
+    let mut creator_enum_variants = Vec::new();
+    let mut creator_enum_len = Vec::new();
+    let mut creator_enum_build = Vec::new();
+    let mut creator_enum_impl_cid = Vec::new();
+
+    for (n, payload, attributes) in members {
+        let n = n.clone();
+        let t = &payload.name;
+        let lt = &payload.lifetime;
+        let doc = &attributes.doc;
+        let creator = Ident::new(&format!("{}Creator", n), Span::call_site());
+
+        // ...len()
+        impl_len.push(quote! {
+            Self::#n(_) => #t::len()
+        });
+
+        // ...bytes()
+        impl_bytes.push(quote! {
+            Self::#n(ref v) => v.bytes()
+        });
+
+        // SerializableMacCommand::cid()
+        impl_cid.push(quote! {
+            Self::#n(_) => #t::cid()
+        });
+
+        // SerializableMacCommand::next()
+        impl_iter_next.push(quote! {
+            if data[0] == #t::cid() && data.len() >= #t::len() {
+                self.index = self.index + #t::len() + 1;
+                Some(#handler::#n(#t::new_from_raw(&data[1..1 + #t::len()])))
+            } else
+        });
+
+        // CommandCreator enum
+        creator_enum_variants.push(quote! {
+            #n(#creator)
+        });
+
+        creator_enum_len.push(quote! {
+            Self::#n(c) => c.len()
+        });
+        creator_enum_build.push(quote! {
+            Self::#n(ref c) => c.build()
+        });
+        creator_enum_impl_cid.push(quote! {
+            Self::#n(c) => c.cid()
+        });
+
+        // Generate definition and common implementation for payloads
+        if let Some((cid, len)) = attributes.attrs {
+            let payload_creator = Ident::new(&format!("{}Creator", n), Span::call_site());
+
+            // Generate common code used by zero and non-zero length payloads
+            let common = quote! {
+                /// Payload CID.
+                pub const fn cid() -> u8 {
+                    #cid
+                }
+                /// Length of payload without the CID.
+                pub const fn len() -> usize {
+                    #len
+                }
+            };
+
+            // Payloads with len > 0 (which have lifetime)
+            if let Some(lt) = lt {
+                payload_struct_impls.push(quote! {
+                    #[derive(Debug, PartialEq, Eq)]
+                    #[cfg_attr(feature = "defmt-03", derive(defmt::Format))]
+                    #( #doc )*
+                    pub struct #t <#lt> (pub(crate) &#lt [u8]);
+
+                    impl<#lt> #t<#lt> {
+                        /// Creates a new instance of the MAC command if there is enought data.
+                        pub fn new(data: &#lt [u8]) -> Result<#t<#lt>, Error> {
+                            if data.len() != #len {
+                                Err(Error::BufferTooShort)
```

Comment:
could we provide self.cid() here to help with possible debug?

Reply:
Yes.. and that error enum contains unused stuff as well.. ;)